### PR TITLE
feat: Provide Azure Application Insights Scaler

### LINKED ISSUE #1965: Problem with fetching Azure custom metric through Azure Monitor scaler
### Discussed in https://github.com/kedacore/keda/discussions/1946

<div type='discussions-op-text'>

<sup>Originally posted by **dymitrs** July 10, 2021</sup>
Hi all, 
It seems that the Monitor scaler does not work with some custom metrics from Application Insights.
According to docs and the PR, Custom Metrics are supported after those changes https://github.com/kedacore/keda/pull/1883.

```
  triggers:
  - type: azure-monitor
    metadata:
      resourceURI: Microsoft.Insights/Components/app-insights
      subscriptionId: <...>
      tenantId: <...>
      resourceGroupName: <...>
      metricName: customMetrics/<metric name>
      metricNamespace: azure.applicationinsights
      metricAggregationInterval: "0:3:0"
      metricAggregationType: Count
      targetValue: "{{ .Values.scale_settings.target_value }}"
```
      
The autoscaler works fine for "default" metrics but for my custom one I'm getting this error and it fails - log from Keda operator pod:

```
ERROR   azure_monitor_scaler    error getting azure monitor metric      {"error": "error getting azure monitor metric customMetrics/<metric name>: got metric result for Microsoft.Insights/customMet
rics/<metric name> and aggregate type COUNT without timeseries"}
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).isScaledObjectActive
        /workspace/pkg/scaling/scale_handler.go:216
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).checkScalers
```
</div>

**Expected behavior:** 
Metrics are read from Azure.

**Actual behavior:**
Error getting metric.

**Steps to reproduce:**
1. Setup Monitor scaler.
2. Fetch custom metrics from Application Insights.

**KEDA Version:**
2.3.0

### PR DESCRIPTION
**Pull request by markrzasa**
Signed-off-by: Mark Rzasa <mark.rzasa@gmail.com>

This pull request adds an Azure Application Insights scaler. This is related to this issue:

kedacore/keda#1965

This pull request is marked as a draft because an app insights scaler may not be necessary to resolve issue 1965. Alternatively, the Azure Log Analytics scaler could be used to read app insights metrics from the AppMetrics table.

Here is the keda-docs PR: https://github.com/kedacore/keda-docs/pull/638

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Changelog has been updated

Fixes #1965

### CODE REVIEW

**Comment by tomkerkhove** [comment]
@markrzasa We're preparing a release for tomorrow. Do you think you can complete the PR before tomorrow in the late morning CET?

**Comment by zroubalik** [comment]
And I think that we need to have these env values defined in the GH secrets.
```
TEST_APP_INSIGHTS_APP_ID
TEST_APP_INSIGHTS_INSTRUMENTATION_KEY
```

**Comment by tomkerkhove** [comment]
> And I think that we need to have these env values defined in the GH secrets.
> 
> ```
> TEST_APP_INSIGHTS_APP_ID
> TEST_APP_INSIGHTS_INSTRUMENTATION_KEY
> ```

Adding now in GitHub as secrets

**Comment by tomkerkhove** [comment]
Before I do so, what identity is it using? The same as other Azure resources? If so, then I need @ahmelsayed to create an AI instance

**Comment by markrzasa** [comment]
The code should match the documentation now.

File: pkg/scalers/azure/azure_app_insights.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+		} else {
+			floatVal = val.(float64)
+		}
+	} else {
+		err = fmt.Errorf("metric %s did not containe aggregation type %s", info.MetricID, info.AggregationType)
+	}
+
+	if err != nil {
+		return -1, err
+	}
```
Comment (JorTurFer) [inline pkg/scalers/azure/azure_app_insights.go:85]: Should we avoid these else?
```suggestion
	if val, ok := metric.Value[info.MetricID].(map[string]interface{})[info.AggregationType]; ok {
		if val == nil {
			return -1, fmt.Errorf("metric %s was nil for aggregation type %s", info.MetricID, info.AggregationType)
		} 
		floatVal = val.(float64)
	} else {
		return -1, fmt.Errorf("metric %s did not containe aggregation type %s", info.MetricID, info.AggregationType)
	}
```

File: pkg/scalers/azure/azure_app_insights_test.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+func newMetricValue(f float64) *float64 {
+	return &f
+}
+
+var testExtractAzAppInsightsData = []testExtractAzAppInsightsTestData{
+	{"metric not found", true, -1, mockAppInsightsInfo("avg"), mockAppInsightsMetric("test/test", "avg", newMetricValue(0.0))},
+	{"metric is nil", true, -1, mockAppInsightsInfo("avg"), mockAppInsightsMetric("testns/test", "avg", nil)},
+	{"incorrect aggregation type", true, -1, mockAppInsightsInfo("avg"), mockAppInsightsMetric("testns/test", "max", newMetricValue(0.0))},
+	{"success round down value", false, 5, mockAppInsightsInfo("max"), mockAppInsightsMetric("testns/test", "max", newMetricValue(5.2))},
+	{"success round up value", false, 5, mockAppInsightsInfo("max"), mockAppInsightsMetric("testns/test", "max", newMetricValue(5.5))},
```
Comment (JorTurFer) [inline pkg/scalers/azure/azure_app_insights_test.go:51]: Round down 5.2 is 5 but round up 5.5 should be 6. Could this be a typo?

**Review by JorTurFer — COMMENTED** [review]
Nice job with this scaler! 😃 
Please, add a unit test ([like this](https://github.com/kedacore/keda/blob/main/pkg/scalers/azure_blob_scaler_test.go#L97-L116)) to check that the metric names are generated correctly
Apart from that, some small nits

File: pkg/scalers/azure/azure_app_insights.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+
+	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	appInsightsResource = "https://api.applicationinsights.io"
```
[the remark this answers is not in the document]
Comment (markrzasa) [reply pkg/scalers/azure/azure_app_insights.go:18]: I believe so at least based on [this](https://docs.microsoft.com/en-us/rest/api/application-insights/metrics/get) documentation.

File: pkg/scalers/azure/azure_app_insights.go (hunk 1 above)
Comment (markrzasa) [reply pkg/scalers/azure/azure_app_insights.go:85]: Yes. Thanks.

File: pkg/scalers/azure/azure_app_insights_test.go (hunk 2 above)
Comment (markrzasa) [reply pkg/scalers/azure/azure_app_insights_test.go]: That was actually a bug in the code that extracts the value. Thanks for catching that.

**Comment by markrzasa** [comment]
> 

should be all set now

**Comment by markrzasa** [comment]
> @markrzasa We're preparing a release for tomorrow. Do you think you can complete the PR before tomorrow in the late morning CET?

I believe all comments have been addressed. If everything looks good, please feel free to complete this PR. Otherwise, I'm not in a big rush to get this merged. I can wait until the next release.

File: pkg/scalers/azure/azure_app_insights.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomkerkhove) [reply pkg/scalers/azure/azure_app_insights.go:19]: We can change it later on in case it does not seem to be the case, but good catch indeed!

**Comment by tomkerkhove** [comment]
Thank you for the PR, very smooth!

> Before I do so, what identity is it using? The same as other Azure resources? If so, then I need @ahmelsayed to create an AI instance

In terms of this, can you elaborate for your e2e tests @markrzasa? We'll need to add a test resource but want to understand what identity it is using now.

Can you fix the DCO issues please? You can learn more in our [contribution guide](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#i-didnt-sign-my-commit-now-what).

**Comment by tomkerkhove** [comment]
We are pushing the release to next Monday so that should give us more time if we want

**Comment by markrzasa** [comment]
> Thank you for the PR, very smooth!
> 
> > Before I do so, what identity is it using? The same as other Azure resources? If so, then I need @ahmelsayed to create an AI instance

In order to run the end-to-end app insights test, you'll need a service principal assigned the 'Monitoring Reader' role scoped to your application insights resource. This setup is likely similar to what is in place for the log analytics scaler.

> In terms of this, can you elaborate for your e2e tests @markrzasa? We'll need to add a test resource but want to understand what identity it is using now.
> 
> Can you fix the DCO issues please? You can learn more in our [contribution guide](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#i-didnt-sign-my-commit-now-what).

Yes. I'll fix that.

**Comment by tomkerkhove** [comment]
> And I think that we need to have these env values defined in the GH secrets.
> 
> ```
> TEST_APP_INSIGHTS_APP_ID
> TEST_APP_INSIGHTS_INSTRUMENTATION_KEY
> ```

Added to GitHub repo's secrets.

**Comment by JorTurFer** [comment]
> > And I think that we need to have these env values defined in the GH secrets.
> > ```
> > TEST_APP_INSIGHTS_APP_ID
> > TEST_APP_INSIGHTS_INSTRUMENTATION_KEY
> > ```
> 
> Added to GitHub repo's secrets.

Did you add them also to the main branch?

**Comment by tomkerkhove** [comment]
See https://github.com/kedacore/keda/pull/2568, ignore PredictKube as I created branch off of the last commit when I added them (sorry)

**Comment by zroubalik** [comment]
@markrzasa could you please rebase your PR, so it contains only relevant commits, it is hard to do a review :)

**Comment by markrzasa** [comment]
> @markrzasa could you please rebase your PR, so it contains only relevant commits, it is hard to do a review :)

I don't mind doing that, but I could use a little help. Is there a specific set of steps or some doc that you would recommend to help me rebase?

**Comment by markrzasa** [comment]
> > @markrzasa could you please rebase your PR, so it contains only relevant commits, it is hard to do a review :)
> 
> I don't mind doing that, but I could use a little help. Is there a specific set of steps or some doc that you would recommend to help me rebase?

I rebased the PR. Hopefully, I handled it correctly.

**Comment by markrzasa** [comment]
> /run-e2e azure-app-insights.* **Update:** You can check the progres [here](https://github.com/kedacore/keda/actions/runs/1757847909)

It looks like the test may need a little more time to scale. I've updated the test to wait a little longer.

**Comment by markrzasa** [comment]
I think I've botched this PR. I see a lot of changes in files that have nothing to do with app insights. Should I discard the PR and start a new one?

File: pkg/scalers/azure/azure_app_insights.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+	"math"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+
+	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
```
Comment (zroubalik) [inline pkg/scalers/azure/azure_app_insights.go:15]: minor nit: would you mind moving `logf "sigs.k8s.io/controller-runtime/pkg/log"` to the block above, next to the Azure deps.

File: pkg/scalers/azure/azure_app_insights.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+		autorest.WithPath(info.ApplicationInsightsID),
+		autorest.WithPath("metrics"),
+		autorest.WithPath(info.MetricID),
+		autorest.WithQueryParameters(queryParams),
+		authorizer.WithAuthorization())
+	if err != nil {
+		return -1, err
+	}
+
+	resp, err := autorest.Send(req,
```
[the remark this answers is not in the document]
Comment (zroubalik) [reply pkg/scalers/azure/azure_app_insights.go:129]: Yeah, looking good.

File: pkg/scalers/azure_app_insights_scaler.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+	"github.com/kedacore/keda/v2/pkg/scalers/azure"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+
+	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
+	kedautil "github.com/kedacore/keda/v2/pkg/util"
+	v2beta2 "k8s.io/api/autoscaling/v2beta2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
```
Comment (zroubalik) [inline pkg/scalers/azure_app_insights_scaler.go:18]: nit: would you mind moving this to the block above?

**Review by zroubalik — COMMENTED** [review]
Looking good, just a minor nits

**Comment by markrzasa** [comment]
> here

Looks good. Thanks for re-running that.

File: pkg/scalers/azure/azure_app_insights.go (hunk 4 above)
Comment (markrzasa) [reply pkg/scalers/azure/azure_app_insights.go]: should be all set now

File: pkg/scalers/azure_app_insights_scaler.go (hunk 6 above)
Comment (markrzasa) [reply pkg/scalers/azure_app_insights_scaler.go]: should be all set now

File: pkg/scalers/azure_app_insights_scaler.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+	"github.com/kedacore/keda/v2/pkg/scalers/azure"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
+	kedautil "github.com/kedacore/keda/v2/pkg/util"
+	v2beta2 "k8s.io/api/autoscaling/v2beta2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (zroubalik) [inline pkg/scalers/azure_app_insights_scaler.go:18]:
```suggestion
	v2beta2 "k8s.io/api/autoscaling/v2beta2"
	"k8s.io/apimachinery/pkg/api/resource"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/labels"
	"k8s.io/metrics/pkg/apis/external_metrics"
	logf "sigs.k8s.io/controller-runtime/pkg/log"

	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
	"github.com/kedacore/keda/v2/pkg/scalers/azure"
	kedautil "github.com/kedacore/keda/v2/pkg/util"
```

We want to have KEDA-related deps in a separate block. Thanks :)

**Review by zroubalik — CHANGES_REQUESTED** [review]

File: pkg/scalers/azure_app_insights_scaler.go (hunk 7 above)
Comment (zroubalik) [reply pkg/scalers/azure_app_insights_scaler.go:18]: @markrzasa one nit pls :)

Comment (markrzasa) [reply pkg/scalers/azure_app_insights_scaler.go]: Got it. Should be all set now.

**Review by zroubalik — APPROVED** [review]
LGTM

**Review by JorTurFer — APPROVED** [review]

**Comment by tomkerkhove** [comment]
Thanks a ton @markrzasa, I love this one!

**Pull request merged**